jfix-stdlib-ratelimiter/src/test/kotlin/ru/fix/stdlib/ratelimiter/SuspendableRateLimitedDispatcherTest.kt

```diff
@@ -0,0 +1,390 @@
+        await().atMost(Duration.ofSeconds(1)).until {
+            trackableDispatcher.isSubmittedTaskInvoked(7..10)
+        }
+        reporter.buildReportAndReset().assertSoftly {
+            it.getMetric("active_async_operations") shouldBe 2
+        }
+
+        await().atMost(Duration.ofSeconds(3)).until {
+            trackableDispatcher.isSubmittedTaskFinished(1..10)
+        }
+        reporter.buildReportAndReset().assertSoftly {
+            it.getMetric("active_async_operations") shouldBe 0
+        }
+
+        trackableDispatcher.completeAllAndClose()
+    }
+
+    /**
+     * When dispatcher closingTimeout is enough for pending tasks to complete
+     * such tasks will complete normally
+     */
+    @Test
+    @Timeout(10, unit = TimeUnit.SECONDS)
+    fun `on shutdown fast tasks complete normally`() = runBlocking {
+
+        val dispatch = createDispatcher(closingTimeout = 5_000)
+
+        val futures = List(3) {
+            dispatch.compose {
+                completedFuture(true)
+            }
+        }
+        dispatch.close()
+
+        futures.forEach { future: CompletableFuture<*> ->
+            future.isDone.shouldBeTrue()
+            future.isCompletedExceptionally.shouldBeFalse()
+        }
+    }
+
+    @Test
+    @Timeout(5, unit = TimeUnit.SECONDS)
+    fun `on shutdown slow tasks complete exceptionally`() = runBlocking {
+        val dispatch = createDispatcher(closingTimeout = 0)
+
+        val deferreds = ArrayList<Deferred<*>>()
+        for (i in 1..3) {
+            deferreds.add(
+                    scope.async {
+                        dispatch.compose {
+                            sleep(3000)
+                            completedFuture(true)
+                        }
+                    }
+            )
+        }
+
+        val exc = shouldThrow<RejectedExecutionException> {
+            dispatch.close()
+        }
+        exc.message!!.contains("timeout while waiting for coroutines finishing").shouldBeTrue()
+    }
+
+    @Test
+    fun `task, submitted in closed dispatcher, is rejected with exception`() = runBlocking {
+        val dispatcher = createDispatcher()
+        dispatcher.close()
+
+        val t = shouldThrow<Exception> {
+            dispatcher.compose {
+                completedFuture(true)
+            }.await()
+        }
+        t.shouldBeInstanceOf<RejectedExecutionException>()
+        t.message!!.contains("TERMINATE")
+        return@runBlocking
+    }
+
+    @Test
+    fun `invoke suspend function wrapped to completable future`() {
+        val dispatcher = createDispatcher()
+
+        val suspendFunctionInvoked = AtomicBoolean(false)
+
+        suspend fun mySuspendFunction(): String{
+            delay(500)
+            suspendFunctionInvoked.set(true)
+            return "suspend-function-result"
+        }
+
+        runBlocking {
+            val result = dispatcher.compose { future { mySuspendFunction() } }.await()
```

Comment:
In context of https://github.com/ru-fix/jfix-stdlib/pull/76#discussion_r1139945474 - I am sure that if we have suspendable function which must be rate limited, we will not wrap it in completable future. We will just do `dispatcher.compose { mySuspendFunction() }`